### DISCUSSION TITLE
How to get the centroid of a collection?

### DISCUSSION THREAD

**Discussion by alanrivers** [milvus-io/milvus#35689] [Q&A and General discussion]
Hi there,

I would like to get the centroid of a collection. And I find this discussion: https://github.com/milvus-io/milvus/discussions/4577, but that seems to work only on v1.x. I wonder how I can get one on v2.x. 

For more information, we have several existing collections representing different clusters. For each incoming vector, we would like to calculate the distance between that vector and centroids of collections ourselves to pick up the nearest collection and add that vector to it.

**Comment by yhmo** [comment]
Currently, v2.x doesn't have such interfaces to get centroids of IVF index.

In fact, milvus manages data by segment, each collection contains multiple partitions(by default, only one partition for one collection), each partition contains multiple segments(the number of segments depends on the data size and compaction machinery).
When you create an IVF index for the collection's vector field, milvus will build an **independent** index for each segment. If the collection has only one segment, the centroids of this segment are the centroids of this collection. However, if the collections have multiple segments, it is difficult to determine which collection is the nearest even if we can get centroids of all the segments.

**Reply by alanrivers** [reply]
Could I get it simply by vector computation:  _sum(centroid * {size of segment}) / {size of collection}_?

**Reply by yhmo** [reply]
Let's say there is a collection A with one segment, and collection B with one segment. Now we get a new vector, we calculate the distance between the vector and the centroids of A and B. It is near to collection A, then we insert it into A. The new inserted vectors might generate a new segment in A, the A gets bigger and hence easily near to new vectors, and gets bigger and bigger... I think it is unexpected.

Before insert:
![centroids_1 drawio (1)](https://github.com/user-attachments/assets/6c4600bf-2319-4288-b10f-8b6db6b2d5cf)

After insert:
![centroids_2 drawio (1)](https://github.com/user-attachments/assets/85754aed-463e-4031-9de0-00f4c6fa6718)

**Reply by alanrivers** [reply]
Would it be possible if the centroid of a collection were calculated once and kept fixed after all labeled vectors were inserted. And the incoming vectors just go in without recalculating the centroids.

**Reply by yhmo** [reply]
Index is based on segment, not on the entire collection. Old segments are merged into new segments. A new segment will get new centroids. No constant centroids. But if the get_centroid() interface is provided, I suppose it returns a map structure, key is segment ID, value is a list of centroids. Then you can get the centroids at a moment and always use these centroids to calculate distance no matter the segments of the collection have been changed/merged.

**Reply by alanrivers** [reply]
Thanks. That might be feasible. Are there any plans about supporting similar issues like this?

**Reply by yhmo** [reply]
Depends on @liliu-z 
In my opinion, it is a valuable function. The interface ought to be a general interface to get detailed information from index, not a specific interface for IVF index. The structure returned by this interface should be carefully designed.

**Comment by liliu-z** [comment]
Hi @alanrivers, we kept thinking about similar features for a while. Can I ask your use case why you need this cluster feature to help us define this requirement more clear?

Thanks

**Reply by alanrivers** [reply]
My goal is to figure out which exisiting collection (we regard each collection as a cluster, and the each collection has already contained prelabeled vectors, which means we predefine clusters using collections) the incoming vector belongs to. And my idea is to get the centroids of all collections by db. So I can compute the distance and choose which collection to insert the incoming vector myself.

**Reply by liliu-z** [reply]
Can I ask why we need to divide vectors? Is that for search acceleration?

**Reply by alanrivers** [reply]
I have several predefined categories with prelabeled vectors, and want to classify the incoming vectors and it's quite natural for vectors of the same category to be put in the same storage unit (maybe a collection). After doing so, I may easily check out statistics of that category by collection apis (eg. to get the current size of each category, to know whether a specific vector is in that category).
